ga4gh/backend.py

```diff
@@ -391,26 +392,150 @@ def _topLevelObjectGenerator(self, request, numObjects, getByIndexMethod):
                 nextPageToken = str(currentIndex)
             yield object_.toProtocolElement(), nextPageToken
 
-    def _objectListGenerator(self, request, objectList):
+    def _filterReadGroupsByBioSampleId(self, bioSampleId, readGroupSet):
         """
-        Returns a generator over the objects in the specified list using
-        _topLevelObjectGenerator to generate page tokens.
+        Accepts a request object containing a biosampleId and
+        removes read groups that don't match the given id.
+        :param str bioSampleId:
+        :param readGroupSet:
+        :return: readGroupSet
         """
-        return self._topLevelObjectGenerator(
-            request, len(objectList), lambda index: objectList[index])
+        newRg = []
+        for rg in readGroupSet.readGroups:
+            if bioSampleId == rg.bioSampleId:
+                newRg.append(rg)
+        readGroupSet.readGroups = newRg
+        return readGroupSet
 
-    def _singleObjectGenerator(self, datamodelObject):
+    def _readGroupSetObjectGenerator(
+            self, request, numObjects, getByIndexMethod, recur=True):
         """
-        Returns a generator suitable for a search method in which the
-        result set is a single object.
+        Returns read group sets that match the given request.
+        Filters read groups by a request's bioSampleId. In order
+        to properly set the nextPageToken the function searches
+        ahead to the next page to see if there are results using
+        a recursive call.
         """
-        yield (datamodelObject.toProtocolElement(), None)
+        currentIndex = 0
+        if request.pageToken is not None:
+            currentIndex, = _parsePageToken(request.pageToken, 1)
+        while currentIndex < numObjects:
+            readGroupSet = getByIndexMethod(currentIndex).toProtocolElement()
+            currentIndex += 1
+            nextPageToken = None
+            if currentIndex < numObjects and recur:
+                request.pageToken = str(currentIndex)
+                for rgs, pageToken in self._readGroupSetObjectGenerator(
+                        request, numObjects, getByIndexMethod, False):
+                    nextPageToken = str(currentIndex)
+                    break
+            if (request.name == "" or request.name) and request.bioSampleId:
+                if request.name == readGroupSet.name:
+                    yield self._filterReadGroupsByBioSampleId(
```

Comment:
Previously read groups and call sets were generating single objects when making a request by name and sent a null pageToken. This was based on the assumption that names were singular. This more general approach uses a look-ahead to determine if the request would generate another object, and if not send a null pageToken.

When a request is made the readGroupSet as selected by the pageToken is being examined. We iterate the currentIndex and perform the same search again having iterated our pageToken. If there are results in the lookahead the nextPageToken is set to the given index.

The schema specifies that `SearchReadGroupSetsResponse` will include only readgroupsets with readgroups that match the bioSampleId, and it will only return readgroups that match the bioSampleId. This means we need to remove non-matching readgroups from a readgroupset if they don't match our request. This is what `filterReadGroupsByBioSampleId` should do.

First, we look at the request to see if it has our query fields present. If both are present we return matches on name, filtered by bioSampleId. If only a name is present we match on name. In the case that only a biosampleId is provided we check to see if the filter leaves behind any readgroups in the readgroupset, and if so returned that filtered object. In the final case, when neither of our query fields are present, simply yield the readgroupset.